tower/src/handle_error/mod.rs

```diff
@@ -0,0 +1,274 @@
+// TODO(david): write docs if we go forward with this
+#![allow(missing_docs)]
+
+use futures_core::ready;
+use pin_project::pin_project;
+use std::{
+    convert::Infallible,
+    fmt,
+    future::Future,
+    marker::PhantomData,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+pub struct HandleErrorLayer<F, R> {
+    f: F,
+    _request: PhantomData<fn() -> R>,
+}
+
+impl<F, R> Clone for HandleErrorLayer<F, R>
+where
+    F: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            f: self.f.clone(),
+            _request: PhantomData,
+        }
+    }
+}
+
+impl<F, R> fmt::Debug for HandleErrorLayer<F, R> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("HandleErrorLayer")
+            .field("f", &format_args!("{}", std::any::type_name::<F>()))
+            .finish()
+    }
+}
+
+impl<F, R> HandleErrorLayer<F, R> {
+    pub fn new(f: F) -> Self {
+        HandleErrorLayer {
+            f,
+            _request: PhantomData,
+        }
+    }
+}
+
+impl<S, R, F> Layer<S> for HandleErrorLayer<F, R>
+where
+    F: Clone,
+    S: Service<R>,
+{
+    type Service = HandleError<S, R, F>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        HandleError::new(inner, self.f.clone())
+    }
+}
+
+#[derive(Clone)]
+pub struct HandleError<S, R, F>
+where
+    S: Service<R>,
+{
+    inner: S,
+    f: F,
+    poll_ready_error: Option<S::Error>,
+}
+
+impl<S, R, F> fmt::Debug for HandleError<S, R, F>
+where
+    S: Service<R> + fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("HandleError")
+            .field("inner", &self.inner)
+            .field("f", &format_args!("{}", std::any::type_name::<F>()))
+            .finish()
+    }
+}
+
+impl<S, R, F> HandleError<S, R, F>
+where
+    S: Service<R>,
+{
+    pub fn new(inner: S, f: F) -> Self
+    where
+        S: Service<R>,
+    {
+        Self {
+            inner,
+            f,
+            poll_ready_error: None,
+        }
+    }
+
+    pub fn layer(f: F) -> HandleErrorLayer<F, R> {
+        HandleErrorLayer::new(f)
+    }
+}
+
+impl<S, R, F> Service<R> for HandleError<S, R, F>
+where
+    S: Service<R>,
+    F: FnOnce(S::Error) -> S::Response + Clone,
+{
+    type Response = S::Response;
+    type Error = Infallible;
+    type Future = ResponseFuture<S::Future, F, S::Error>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        if let Err(err) = ready!(self.inner.poll_ready(cx)) {
```

Comment:
Doesn't this break the service contract?

> If Poll::Ready(Err(_)) is returned, the service is no longer able to service requests and the caller should discard the service instance.

Reply:
Hm yeah I suppose it does :(

Reply:
Yeah, I think errors from `poll_ready` should be propagated --- errors from `call`, on the other hand, _don't_ indicate that the service is unusable, and should be handled.

Reply:
I suppose that means this general strategy of changing the error type to `Infallible` and always producing a response isn't possible 🤔

Reply:
You could maybe have a service whose error type is `Infallible` if it _also_ included a `MakeService` and some mechanism for making new services and driving them for readiness when the inner service fails. You would probably also need to include a backoff strategy for when making a new service fails.

Potentially, there could be a helper to combine a `MakeService`/reconnect strategy with a `HandleError` middleware, and then map all errors to `Infallible`, since we would then know that both `poll_ready` _and_ `call` errors would be handled nicely in that case.